### DISCUSSION TITLE
psycopg.ProgrammingError: cannot adapt type 'RepeatedScalarContainer' using placeholder '%s' (format: AUTO)

### DISCUSSION THREAD

**Discussion by bryanhughes** [psycopg/psycopg#277] [Q&A]
Hello,

I am new to psycopg and python (expertise is Erlang, Java, and Go), I am stuck trying to insert into a column defined as:

`my_array             integer[]  NOT NULL `

When I try to insert into the table, I am getting

`psycopg.ProgrammingError: cannot adapt type 'RepeatedScalarContainer' using placeholder '%s' (format: AUTO)`

I googled around and only found an answer to use the `cur.mogrify()` but that has been removed from psycopg 3

Any guidance on how to address this issue. Using scalar arrays in Postgres is not unusual.

Thanks!
Bryan

**Comment by dvarrazzo** [comment]
Hi Brian. Erlang is cool! ✌️

I am not familiar with the `RepeatedScalarContainer` I understand from quick googling that they are related to proto buffers.

The right way to adapt them is to create a `psycopg.adapter.Dumper` subclass, implementing the `dump()` method, which would take one such object and return a bytes representation of the data for Postgres.

Because the container is likely a generic object, it wouldn't map easily to a specific postgres type. As such you are better off leaving the dumper oid to 0 and only implement text dumping.

All the details should be in the [adaptation customisation docs](https://www.psycopg.org/psycopg3/docs/advanced/adapt.html).

However, if that container looks like a python list, you might get away with just registering the RepeatedScalarContainer to the list dumper, or by subclassing, or copying and editing the LiatDumper class.

**Reply by bryanhughes** [reply]
Thanks for the quick reply! I should have caught that...wasnt sure on what side of the fence this was happening on. I will try your suggestion and get back to you.

Cheers!
